Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -16,130 +16,313 @@ import SymbolKit
+        
+        // This code get called for every symbol with a type signature, so it needs to be fast.
+        // Because all the characters that need to be identified and processed are UTF-8 code units, this implementation works solely on UTF-8 code units.
+        var accumulated = ContiguousArray<UTF8.CodeUnit>()
+        // Reserve some temporary space to work with. This avoids reallocations for most declarations.
+        // The final string will make it's own copy, so this temporary memory is only used until the end of this scope.
+        accumulated.reserveCapacity(128)
         
+        // Iterating over the declaration fragments to accumulate their spelling and to identify places that need to apply syntactic sugar.
+        var markers = ContiguousArray<Int>()
         for fragment in fragments {
-            switch fragment.kind {
-            case .identifier where fragment.preciseIdentifier != nil,
-                 .typeIdentifier,
-                 .text:
-                accumulated += fragment.spelling
-
-            default:
-                continue
+            let preciseIdentifier = fragment.preciseIdentifier
+            if isSwift {
+                // Check if this fragment is a spelled out Swift array, optional, or dictionary.
+                switch preciseIdentifier {
+                case "s:Sa", // Swift.Array
+                     "s:SD", // Swift.Dictionary
+                     "s:Sq": // Swift.Optional
+                    assert(fragment.spelling == "Array" || fragment.spelling == "Dictionary" || fragment.spelling == "Optional", """
+                    Unexpected spelling '\(fragment.spelling)' for Array/Dictionary/Optional fragment in declaration; \(fragments.map(\.spelling).joined())
+                    """)
+                    
+                    // Create a new marker at this location and insert is at the beginning since `withSwiftSyntacticSugar(markers:)`
+                    // will iterate the markers from the end to the start.
+                    markers.insert(accumulated.count, at: 0)
+                    
+                    // Since `withSwiftSyntacticSugar(markers:)` below will remove this spelling, only collect the first character.
+                    // The first character ("A", "D", or "O") is sufficient to identify which type of sugar to apply.
+                    accumulated.append(fragment.spelling.utf8.first!)
+                    continue
+                    
+                default:
+                    break
+                }
+                
+                switch fragment.kind {
+                case .typeIdentifier:
+                    // Accumulate all of the identifier tokens' spelling.
+                    accumulated.append(contentsOf: fragment.spelling.utf8)
+                    
+                case .text: // In Swift, only `text` tokens contains whitespace
```

Comment:
What's an example of a value that `text.spelling` can have? Can you share an example declaration fragment for this case?

Reply:
I added a comment in https://github.com/swiftlang/swift-docc/pull/1047/commits/4a6b326f068ded10576d57812ddd4668cfb9c18e with some example text tokens and how each contribute to a type spelling.